cryptogenotyper v1.0 release

kbessonov1984 opened the pull request:
The `CryptoGenotyper` is a fast and reproducible tool that can be used to classify the genotype of *Cryptosporidium* samples by directly analyzing the DNA electropherogram files (.ab1) that correspond to two of its characteristic gene markers: SSU rRNA and gp60.

This is the first official Galaxy release of the tool. I had tested it in `planemo v0.70.0` and did not find any issues.

Partially linked to issue [https://github.com/christineyanta/crypto_typer/issues/1](https://github.com/christineyanta/crypto_typer/issues/1)

apetkau commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,135 @@
+  <description> CryptoGenotyper is a standalone tool to analyze <i>Cryptosporidium</i> Sanger sequencing data and classify the species and subtype based on SSU rRNA and <i>gp60</i> gene markers, respectively. </description>
+  <requirements>
+    <requirement type="package" version ="1.0">cryptogenotyper</requirement>
+  </requirements>
```
Could you add in a `<version_command>` section like for [staramr](https://github.com/phac-nml/galaxy_tools/blob/master/tools/staramr/staramr_search.xml#L9). This would be the command to run to print the underlying tool version.

kbessonov1984 replied:
I've added it. Thank you

apetkau commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,135 @@
+     <![CDATA[
+
+     #set $ref_file=''
+     #set $ref_name=''
```
Is `$ref_name` used in the rest of the code at all?

kbessonov1984 replied:
I removed it. I am not sure why it was added in the first place by Christine. Looks like an old artifact

apetkau commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,135 @@
+        <param name="seqtype" type="select" label="Type of Sequences">
+            <option value="forward">Forward Only</option>
+            <option value="reverse">Reverse Only</option>
+            <option selected="true" value="contig">Contig</option>
```
I'm not really familiar with this type of data, but is the appropriate word `contig` here or is it `read`? Here you use **Contig**, but elsewhere when talking about paired data you use **Read**.

kbessonov1984 replied:
It is indeed a good observation. I had added a clarification in the description of the tool to indicate that contig mode related to forward and reverse paired reads. Apparently in the context of Cryptosporidium Sanger sequencing this terminology is used. Also in the manuscript it this paired-end mode is also referred as contig mode.  From what I understand given an amplicon it is possible to type Cryptosporidium based on forward primer (forward read) or reverse read (reverse primer) or using both forward and reverse reads (contig mode)

apetkau replied:
Okay, thanks. That makes sense. I just wanted to make sure the terminology was consistent.

apetkau reviewed (changes requested):
Overall this looks great, and seems to all run for me on my computer. In addition to the few comments below I have some other requests:

1. Could you add a `.shed.yml` file specifying where this tool belongs in the Galaxy Toolshed? You can look at what was done with the [sistr](https://github.com/phac-nml/galaxy_tools/blob/cryptogenotyper/tools/sistr_cmd/.shed.yml) tool.
2. When running the planemo linter `planemo l` I get the following errors:

    ```
    .. ERROR: Invalid XML found in file: cryptogenotyper.xml. Errors [/tmp/tmpyy2h5q0f:2:0:ERROR:SCHEMASV:SCHEMAV_CVC_TYPE_3_1_2: Element 'description': Element content is not allowed, because the type definition is simple.
    /tmp/tmpyy2h5q0f:84:0:ERROR:SCHEMASV:SCHEMAV_CVC_COMPLEX_TYPE_3_2_1: Element 'param', attribute 'type': The attribute 'type' is not allowed.
    /tmp/tmpyy2h5q0f:84:0:ERROR:SCHEMASV:SCHEMAV_CVC_COMPLEX_TYPE_3_2_1: Element 'param', attribute 'format': The attribute 'format' is not allowed.
    /tmp/tmpyy2h5q0f:94:0:ERROR:SCHEMASV:SCHEMAV_CVC_COMPLEX_TYPE_3_2_1: Element 'param', attribute 'type': The attribute 'type' is not allowed.
    /tmp/tmpyy2h5q0f:94:0:ERROR:SCHEMASV:SCHEMAV_CVC_COMPLEX_TYPE_3_2_1: Element 'param', attribute 'foramt': The attribute 'foramt' is not allowed.]
    Failed linting
    ```

    Could you look into why these are happening?

Thanks so much.

kbessonov1984 commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,135 @@
+<tool id="CryptoGenotyper" name = "CryptoGenotyper" version = "1.0">
```
Sure, I like to use this macros feature. Thank you

kbessonov1984 commented:
I've added `.shed.yml` file. I've  had it but forgot to run `git add .shed.yml`.
I also found source of lint error. Just needed to remove some data specifications in the ` <param name="abi_input" ....>` as they were reduntant

apetkau commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,141 @@
+   <token name="@VERSION@">1.0</token>
+  </macros>
+  <requirements>
+    <requirement type="package" version ="1.0">cryptogenotyper</requirement>
```
Could you replace this with `@VERSION@` too?

kbessonov1984 replied:
Sure, I forgot about this tag too. Thank you

apetkau reviewed (changes requested):
Thanks for addressing these issues. Just one more small change.

apetkau commented:
Looks good. Thanks @kbessonov1984 . I will merge.

bgruening commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,141 @@
+     #if $primers["seqtype"] == "contig"
+        ln -s "${$primers.abi_input["forward"]}" "${primers.abi_input.name}_forward.ab1" &&
+        ln -s "${$primers.abi_input["reverse"]}" "${primers.abi_input.name}_reverse.ab1" &&
+        cryptogenotyper -i "." -m "$marker" -t "$primers.seqtype" -f "forward" -r "reverse"
```
@kbessonov1984 please use single quotes

kbessonov1984 replied:
Done

bgruening commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,141 @@
+     #set $ref_file=''
+
+     #if $reference.ref == "no"
+        ln -s "${reference.db}" "${reference.db.name}" &&
```
single quotes please

kbessonov1984 replied:
Done

bgruening commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,141 @@
+        ln -s "${$primers.abi_input["reverse"]}" "${primers.abi_input.name}_reverse.ab1" &&
+        cryptogenotyper -i "." -m "$marker" -t "$primers.seqtype" -f "forward" -r "reverse"
+        #if $reference.ref == "no"
+          --databasefile $ref_file
```
please add quotes

bgruening commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,141 @@
+        ln -s "${primers.abi_input}" "${primers.abi_input.element_identifier}" &&
+        cryptogenotyper -i "./${primers.abi_input.element_identifier}" -m "$marker" -t "$primers.seqtype" -f "" -r ""
+        #if $reference.ref == "no"
+          --databasefile $ref_file
```
quotes :)

kbessonov1984 replied:
Done. Thank you, I've read that single quotes from Cheetah variables are more secure

bgruening commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,141 @@
+     #if $outputheader == "no"
+           --noheaderline
+     #end if
+     -o "result";
```
you can remove the trailing `;`

kbessonov1984 replied:
Done

bgruening commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,141 @@
+        <when value="yes">
+        </when>
+        <when value="no">
+            <param name="db" type="data" format="fasta" label="Reference Database File:" help=".fa fasta file type"/>
```
you could remove the entire conditional and make this parameter optional="true"

kbessonov1984 replied:
I had tried with the optional parameter option, but then I need `<param name="db" type="data" format="fasta" label="Reference Database File:" help=".fa fasta file type"/>` to be executed when user selects No (meaning that user wants to user his/her own custom database of reference alleles). I know it is a bit ugly from the code perspective, but I do not know how to make it work when decorative "Yes" option is selected as a filler. Thus I left it as it is.

bgruening replied:
Sorry, that I do not get. Can you not remove the entire condition and just put `<param name="db" type="data" optional="true" format="fasta" label="Reference Database File:" help=".fa fasta file type"/>`?

I'm not worried about the code so much, but the UX should be better because you need on less click :)

kbessonov1984 replied:
I've tried to make it optional but then it crashes as then it is possible to launch the program when there is no custom database specified. Otherwise if will make it non-optional, user will see a red section highlighted preventing the launch. By default user does not need to specify a database as the default one is used instead.
Maybe we can make it such that this custom field is always visible but is optional so there is no need to select Yes or No. If it is defined, then it is used.

bgruening replied:
Oh I think I understand now. What you need to do is add something like this to cheetah:

```
#if $db:
   do something when a file was selected
#end if
```

kbessonov1984 replied:
Yes, let's make it always visible field and then if it is defined, execute that parameter. There will be 1 less click in case user would decide to provide a custom DB. We are now releasing a version +galaxy 0 with all changes. Many thanks!

bgruening commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,141 @@
+          <param name="abi_input" type="data_collection" collection_type="paired" format="ab1" label="Paired Sequencing File(s)" help=".ab1 file type"/>
+        </when>
+        <when value="forward">
+          <param name="abi_input" type="data" format="ab1" label="Forward Sequencing File(s)" help=".ab1 file type"/>
```
The help seems to be redundant as Galaxy will tell you the filetype. A more detailed help about the file would be good or remove the help.

kbessonov1984 replied:
Yes, I see that when there are no input datasets, Galaxy provides a hint, so help is redundant. I had removed it

bgruening commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,141 @@
+  </outputs>
+
+  <tests>
+    <test>
```
can you please add `expected_num_output="2"`

kbessonov1984 replied:
Implemented. Great security feature to add. Did not know about it.

bgruening commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,141 @@
+        #end if
+     #else
+        ln -s "${primers.abi_input}" "${primers.abi_input.element_identifier}" &&
+        cryptogenotyper -i "./${primers.abi_input.element_identifier}" -m "$marker" -t "$primers.seqtype" -f "" -r ""
```
what does -f and -r is doing?

kbessonov1984 replied:
In this case these parameters are redundant. They are used to filter input files based on names of the primers, but since in Galaxy will work with maximum a pair of files at a time, then it is redundant indeed. I had removed it

bgruening commented on tools/cryptogenotyper/cryptogenotyper.xml:
```diff
@@ -0,0 +1,141 @@
+<tool id="CryptoGenotyper" name = "CryptoGenotyper" version="@VERSION@">
+  <description>
+    CryptoGenotyper is a standalone tool to analyze Cryptosporidium Sanger sequencing data and classify the species and subtype based on SSU rRNA and gp60 gene markers, respectively.
```
This description will be in the tool panel, it seems a little bit too long.

Maybe something like "classify species and subtypes based on SSU rRNA and gp60 on Cryptosporidium Sanger sequencing data"?

kbessonov1984 replied:
Changed to a shorter title to make it look prettier.

bgruening reviewed (commented):
@kbessonov1984 sorry that I missed this in the first place. I added some comments inline.

bgruening commented:
@kbessonov1984 if you want this to be added to EU. You can add it simply to this file: https://github.com/usegalaxy-eu/usegalaxy-eu-tools/blob/master/nml.yaml and it will get installed on each Saturday. Thanks for your work!

kbessonov1984 commented:
@bgruening Great. I had addressed most of the comments. Will re-release package then. To make tool available on EU Galaxy, it is enough to just add tool name to that `nml.yaml` file via a Pull Request?

bgruening commented:
> @bgruening Great. I had addressed most of the comments. Will re-release package then. To make tool available on EU Galaxy, it is enough to just add tool name to that `nml.yaml` file via a Pull Request?

Yes :)
